presto-main/src/main/java/com/facebook/presto/operator/exchange/LocalMergeSinkOperator.java

```diff
@@ -0,0 +1,154 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator.exchange;
+
+import com.facebook.presto.operator.DriverContext;
+import com.facebook.presto.operator.Operator;
+import com.facebook.presto.operator.OperatorContext;
+import com.facebook.presto.operator.OperatorFactory;
+import com.facebook.presto.operator.exchange.LocalMergeExchange.ExchangeBuffer;
+import com.facebook.presto.spi.Page;
+import com.facebook.presto.spi.type.Type;
+import com.facebook.presto.sql.planner.plan.PlanNodeId;
+import com.google.common.util.concurrent.ListenableFuture;
+
+import java.util.List;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+
+public class LocalMergeSinkOperator
+        implements Operator
+{
+    public static class LocalMergeSinkOperatorFactory
+            implements OperatorFactory
+    {
+        private final int operatorId;
+        private final PlanNodeId planNodeId;
+        private final List<Type> types;
+        private final LocalMergeExchange exchange;
+        private final int operatorsCount;
+        private final AtomicInteger operatorIndex = new AtomicInteger();
+        private boolean closed;
+
+        public LocalMergeSinkOperatorFactory(int operatorId, PlanNodeId planNodeId, List<Type> types, LocalMergeExchange exchange, int operatorsCount)
+        {
+            this.operatorId = operatorId;
+            this.planNodeId = requireNonNull(planNodeId, "planNodeId is null");
+            this.types = requireNonNull(types, "types is null");
+            this.exchange = requireNonNull(exchange, "exchange is null");
+            checkArgument(operatorsCount > 0, "operatorsCount must be greater than zero");
+            this.operatorsCount = operatorsCount;
+        }
+
+        @Override
+        public List<Type> getTypes()
+        {
+            return types;
+        }
+
+        @Override
+        public Operator createOperator(DriverContext driverContext)
+        {
+            checkState(!closed, "closed");
+            checkState(operatorIndex.get() < operatorsCount);
```

Comment:
I though about this. But then you should make `MergeExchange` implement `Exchange` what introduces additional complexity.